vm/entrypoint/src/syscalls/host.rs

```diff
@@ -41,57 +42,54 @@ pub extern "C" fn call(
 }
 
 /// Read from host storage at the given address and key.
-///
-/// The output is stored in the `key` pointer.
 #[allow(unused_variables)]
-#[no_mangle]
-pub extern "C" fn read_storage(key: *mut u32) {
+pub fn read_storage(key: &[u32; 8]) -> [u32; 8] {
   #[cfg(target_os = "zkvm")]
   unsafe {
+    let mut result = *key;
     asm!(
         "ecall",
         in("t0") crate::syscalls::HOST_READ,
-        in("a0") key,
-    )
+        in("a0") result.as_mut_ptr(),
+    );
+    return result;
   }
 
   #[cfg(not(target_os = "zkvm"))]
   unreachable!()
 }
 
 /// Write to host storage at the given address and key.
-///
-/// The result status code is stored in the `key` pointer.
 #[allow(unused_variables)]
-#[no_mangle]
-pub extern "C" fn write_storage(key: *mut u32, value: *const u32) {
+pub fn write_storage(key: &[u32; 8], value: &[u32; 8]) -> StorageStatus {
   #[cfg(target_os = "zkvm")]
   unsafe {
+    let status: u32;
     asm!(
         "ecall",
         in("t0") crate::syscalls::HOST_WRITE,
-        in("a0") key,
-        in("a1") value,
-    )
+        in("a0") key.as_ptr(),
+        in("a1") value.as_ptr(),
+        lateout("t0") status,
+    );
+    return status.try_into().unwrap();
   }
 
   #[cfg(not(target_os = "zkvm"))]
   unreachable!()
 }
 
 /// Get the current account balance
-///
-/// The result status code is stored in the `value` pointer.
-#[allow(unused_variables)]
-#[no_mangle]
-pub extern "C" fn get_balance(value: *mut u32) {
+pub fn get_balance() -> u64 {
   #[cfg(target_os = "zkvm")]
   unsafe {
+    let mut balance = std::mem::MaybeUninit::<u64>::uninit();
     asm!(
         "ecall",
         in("t0") crate::syscalls::HOST_GETBALANCE,
-        in("a0") value,
-    )
+        in("a0") balance.as_mut_ptr(),
+    );
+    balance.assume_init()
```

Comment:
possibly dumb question: how do we ensure that the value is read here as LE? I know the value is written that way, but what if the host architecture here is BE?

Reply:
We don't and I think can't ensure this. However, I suppose none of the program <-> VM communication would work if the host machine was BE. For example, if the program wrote a `0x000000CA` into a register, the host would read `0xCA000000` and nothing would work. To support a BE host machine, I think we would need to always go through a LE (program) <-> native (host) conversion (sometimes a noop, if host is also LE), whenever we read and write registers and program memory on the VM side.

Reply:
We can always ensure this if we want to but it's extra work. See e.g. the conversion code here, using `bytemuck`: https://github.com/risc0/risc0/blob/0e6169b2fe652cc3e62927a5347b4af496ad62dd/risc0/zkvm/src/host/server/testutils.rs#L22-L37.

We're not alone in this issue: https://github.com/risc0/risc0/issues/420. I think it's fine for now. I opened #141 so it's documented/searchable. It would be good to point out places in the code that make this assumption, unless you think there are just too many of them.

Reply:
I think it's every place where the VM reads or writes a register or program memory.